edb/edgeql-parser/src/parser.rs

```diff
@@ -0,0 +1,590 @@
+
+                    let cost = error_cost(token_kind);
+                    let error = Error::new(format!("Missing {injection}")).with_span(gap_span);
+                    inject.push_error(error, cost);
+
+                    if inject.error_cost <= ERROR_COST_INJECT_MAX {
+                        // println!("   --> [inject {injection}]");
+
+                        if inject.act(ctx, injection).is_ok() {
+                            // insert into parsers, to retry the original token
+                            parsers.push(inject);
+                        }
+                    }
+                }
+
+                // option 2: skip the token
+
+                let mut skip = parser;
+                let error = Error::new(format!("Unexpected {token}")).with_span(token.span);
+                skip.push_error(error, ERROR_COST_SKIP);
+                if token.kind == Kind::EOF {
+                    // extra penalty
+                    skip.error_cost += ERROR_COST_INJECT_MAX;
+                    skip.can_recover = false;
+                };
+
+                // println!("   --> [skip]");
+
+                // insert into new_parsers, so the token is skipped
+                new_parsers.push(skip);
+            }
+        }
+
+        // has any parser recovered?
+        if new_parsers.len() > 1 {
+            let recovered = new_parsers.iter().position(Parser::has_recovered);
+
+            if let Some(recovered) = recovered {
+                let mut recovered = new_parsers.swap_remove(recovered);
+                recovered.error_cost = 0;
+
+                new_parsers.clear();
+                new_parsers.push(recovered);
+            }
+        }
+
+        // prune: pick only X best parsers
+        if new_parsers.len() > PARSER_COUNT_MAX {
+            new_parsers.sort_by_key(Parser::adjusted_cost);
+            new_parsers.drain(PARSER_COUNT_MAX..);
+        }
+
+        assert!(parsers.is_empty());
+        std::mem::swap(&mut parsers, &mut new_parsers);
+        prev_span = Some(token.span);
+    }
+
+    // there will always be a parser left,
+    // since we always allow a token to be skipped
+    let mut parser = parsers.into_iter().min_by_key(|p| p.error_cost).unwrap();
+    parser.finish();
+
+    let node = if parser.can_recover {
+        Some(&parser.stack_top.value)
+    } else {
+        None
+    };
+    (node, parser.errors)
+}
+
+impl<'s> Context<'s> {
+    fn alloc_terminal(&self, t: Terminal) -> &'_ Terminal {
+        let idx = self.terminal_arena.push(t);
+        &self.terminal_arena[idx]
+    }
+}
+
+fn new_token_for_injection<'a>(kind: &Kind, ctx: &'a Context) -> &'a Terminal {
+    ctx.alloc_terminal(Terminal {
+        kind: kind.clone(),
+        text: kind.text().unwrap_or_default().to_string(),
+        value: match kind {
+            Kind::Keyword(Keyword(kw)) => Some(Value::String(kw.to_string())),
+            Kind::Ident => Some(Value::String("my_name".to_string())),
```

Comment:
Could you make the synthesized identifier name something more distinctive? Maybe something that is actually an invalid identifier. (Contains a backtick?)
Tools will want to be able to detect this case.